mountpoint-s3/src/prefetch.rs

```diff
@@ -285,15 +288,51 @@ where
 
         // [read] will reset these if the reader stops making sequential requests
         self.next_request_offset += size;
-        self.next_request_size = (self.next_request_size * self.inner.config.sequential_prefetch_multiplier)
-            .min(self.inner.config.max_request_size);
+        self.next_request_size = Self::get_next_request_size(
+            self.next_request_offset,
+            self.next_request_size,
+            self.inner.config.sequential_prefetch_multiplier,
+            self.inner.config.max_request_size,
+            self.inner.config.part_size,
+        );
 
         Some(RequestTask {
             total_size: size as usize,
             remaining: size as usize,
             part_queue,
         })
     }
+
+    /// Suggest next request size.
+    /// Normally, next request size is current request size multiply by sequential prefetch multiplier,
+    /// but if the request size is getting bigger than a part size we will try to align it to part boundaries.
+    fn get_next_request_size(
+        next_request_offset: u64,
+        current_request_size: usize,
+        multiplier: usize,
+        max_request_size: usize,
+        part_size: usize,
+    ) -> usize {
+        // calculate next request size
+        let next_request_size = (current_request_size * multiplier).min(max_request_size);
+
+        let offset_in_part = (next_request_offset % part_size as u64) as usize;
+        // if the offset is not at the start of the part we will drain all the bytes from that part first
+        if offset_in_part != 0 {
+            let remaining_in_part = part_size - offset_in_part;
+            return next_request_size.min(remaining_in_part);
+        }
+
+        // if the next request size is smaller than the part size, just return that value
+        if next_request_size < part_size {
+            return next_request_size;
+        }
```

Comment:
The logic will be a bit different because we only do this when the offset is at the start of the part. I wrapped it in `else` block instead, that should make it clearer.